### ISSUE TITLE
Epic: Model Fine-tuning Pipeline

### ISSUE THREAD

**Issue by mikesvoboda** [mikesvoboda/nemotron-v3-home-security-intelligence#3992]
## Overview

Create a pipeline for fine-tuning detection and analysis models on user-specific data. This enables the system to learn the user's environment (recognize family members, regular vehicles, typical patterns) and reduce false positives while improving detection of genuine threats.

## Current Architecture

**Current Models:**

* RT-DETRv2: Pre-trained on COCO, detects 80 generic classes
* Nemotron 70B: General-purpose LLM with security prompt
* Enrichment models: Pre-trained on general datasets

**Limitations:**

* No learning from user feedback
* High false positive rate for known entities
* No environment-specific optimization
* Same thresholds for all cameras/zones

### Fine-tuning Targets

| Model | Fine-tuning Goal | Method |
| -- | -- | -- |
| **RT-DETRv2** | Detect custom objects, reduce FP on known entities | Transfer learning |
| **Face Recognition** | Learn family/known persons | Embedding fine-tuning |
| **Vehicle Recognition** | Learn household vehicles | Embedding fine-tuning |
| **Nemotron** | Learn user's risk preferences | LoRA/QLoRA |

### Data Collection Pipeline

```
User Feedback → Labeling Queue → Annotation UI → Dataset Builder → Training Pipeline
     │                                                                    │
     └─── "Mark as Known" ──→ Known Entity Registry ←──── Fine-tuned Model
```

### New Components

1. **Feedback Collector** (`backend/services/feedback_collector.py`)
   * Collect user annotations (correct/incorrect, known entity)
   * Store feedback with detection metadata
2. **Dataset Builder** (`backend/services/dataset_builder.py`)
   * Build training datasets from feedback
   * Data augmentation
   * Train/val/test splits
3. **Training Pipeline** (`training/`)
   * Model fine-tuning scripts
   * Hyperparameter management
   * Model versioning
4. **Known Entity Registry** (`backend/services/known_entities.py`)
   * Track known persons, vehicles
   * Embedding storage and matching

### Feedback Data Model

```python
# backend/models/feedback.py
class DetectionFeedback(Base):
    __tablename__ = "detection_feedback"
    
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    detection_id: Mapped[UUID] = mapped_column(ForeignKey("detections.id"))
    
    # Feedback type
    feedback_type: Mapped[str]  # "correct", "incorrect", "known_entity", "relabel"
    
    # For relabeling
    original_label: Mapped[str | None]
    corrected_label: Mapped[str | None]
    
    # For known entities
    known_entity_id: Mapped[UUID | None] = mapped_column(ForeignKey("known_entities.id"))
    
    # Metadata
    created_at: Mapped[datetime] = mapped_column(default=func.now())
    user_id: Mapped[UUID | None]
    
    detection: Mapped["Detection"] = relationship()
    known_entity: Mapped["KnownEntity"] = relationship()


class KnownEntity(Base):
    __tablename__ = "known_entities"
    
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    name: Mapped[str] = mapped_column(String(100))
    entity_type: Mapped[str]  # "person", "vehicle", "pet"
    
    # Recognition data
    embeddings: Mapped[list[list[float]]] = mapped_column(JSONB)  # Multiple embeddings per entity
    
    # Metadata
    thumbnail_path: Mapped[str | None]
    notes: Mapped[str | None]
    created_at: Mapped[datetime] = mapped_column(default=func.now())
    
    # Settings
    always_safe: Mapped[bool] = mapped_column(default=False)  # Skip alerts for this entity
    risk_modifier: Mapped[int] = mapped_column(default=0)  # Adjust risk score (-50 to +50)


class TrainingRun(Base):
    __tablename__ = "training_runs"
    
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    model_type: Mapped[str]  # "detector", "face", "vehicle", "llm"
    
    # Dataset info
    dataset_size: Mapped[int]
    train_samples: Mapped[int]
    val_samples: Mapped[int]
    
    # Training config
    config: Mapped[dict] = mapped_column(JSONB)
    
    # Results
    status: Mapped[str]  # "pending", "running", "completed", "failed"
    metrics: Mapped[dict | None] = mapped_column(JSONB)
    model_path: Mapped[str | None]
    
    # Timestamps
    started_at: Mapped[datetime | None]
    completed_at: Mapped[datetime | None]
    created_at: Mapped[datetime] = mapped_column(default=func.now())
```

### Feedback API

```python
# backend/api/routes/feedback.py
router = APIRouter(prefix="/api/feedback", tags=["feedback"])

@router.post("/detections/{detection_id}/correct")
async def mark_detection_correct(
    detection_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Mark a detection as correct (positive training sample)."""
    await feedback_service.add_feedback(
        detection_id=detection_id,
        feedback_type="correct",
    )
    return {"status": "recorded"}

@router.post("/detections/{detection_id}/incorrect")
async def mark_detection_incorrect(
    detection_id: UUID,
    corrected_label: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Mark a detection as incorrect (negative training sample)."""
    await feedback_service.add_feedback(
        detection_id=detection_id,
        feedback_type="incorrect" if not corrected_label else "relabel",
        corrected_label=corrected_label,
    )
    return {"status": "recorded"}

@router.post("/detections/{detection_id}/known-entity")
async def mark_as_known_entity(
    detection_id: UUID,
    request: KnownEntityRequest,
    db: AsyncSession = Depends(get_db),
):
    """Associate detection with a known entity (person/vehicle)."""
    # Create or find known entity
    if request.known_entity_id:
        entity = await db.get(KnownEntity, request.known_entity_id)
    else:
        # Create new known entity
        entity = KnownEntity(
            name=request.name,
            entity_type=request.entity_type,
        )
        db.add(entity)
    
    # Extract and store embedding
    detection = await db.get(Detection, detection_id)
    embedding = await embedding_service.extract(detection.crop_path)
    entity.embeddings.append(embedding)
    
    # Record feedback
    await feedback_service.add_feedback(
        detection_id=detection_id,
        feedback_type="known_entity",
        known_entity_id=entity.id,
    )
    
    await db.commit()
    return KnownEntityResponse.from_orm(entity)

@router.get("/known-entities")
async def list_known_entities(
    entity_type: str | None = None,
) -> list[KnownEntityResponse]:
    """List all known entities."""
    ...

@router.post("/training/start")
async def start_training(
    request: TrainingRequest,
    user: User = Depends(require_admin),
):
    """Start a training run with collected feedback."""
    run = await training_service.start_training(
        model_type=request.model_type,
        config=request.config,
    )
    return TrainingRunResponse.from_orm(run)
```

### Dataset Builder

```python
# backend/services/dataset_builder.py
class DatasetBuilder:
    async def build_detector_dataset(
        self,
        min_samples: int = 100,
    ) -> DetectorDataset:
        """Build dataset for detector fine-tuning."""
        
        # Get positive samples (correct detections)
        positives = await self.feedback_repo.get_feedback(
            feedback_type="correct",
            limit=min_samples * 2,
        )
        
        # Get negative samples (incorrect detections)
        negatives = await self.feedback_repo.get_feedback(
            feedback_type="incorrect",
            limit=min_samples,
        )
        
        # Get relabeled samples
        relabeled = await self.feedback_repo.get_feedback(
            feedback_type="relabel",
            limit=min_samples,
        )
        
        # Build COCO-format dataset
        dataset = COCODataset()
        
        for feedback in positives:
            detection = feedback.detection
            dataset.add_annotation(
                image_path=detection.image_path,
                bbox=detection.bbox,
                label=detection.label,
                is_positive=True,
            )
        
        for feedback in negatives:
            detection = feedback.detection
            dataset.add_hard_negative(
                image_path=detection.image_path,
                bbox=detection.bbox,
                label=detection.label,
            )
        
        for feedback in relabeled:
            detection = feedback.detection
            dataset.add_annotation(
                image_path=detection.image_path,
                bbox=detection.bbox,
                label=feedback.corrected_label,
                is_positive=True,
            )
        
        # Split train/val
        return dataset.split(train_ratio=0.8)
```

### Training Pipeline

```python
# training/detector_finetune.py
import torch
from transformers import RTDetrForObjectDetection

class DetectorFineTuner:
    def __init__(self, base_model: str = "PekingU/rtdetr_r50vd"):
        self.model = RTDetrForObjectDetection.from_pretrained(base_model)
        self.processor = RTDetrImageProcessor.from_pretrained(base_model)
    
    def train(
        self,
        dataset: DetectorDataset,
        output_dir: str,
        epochs: int = 10,
        learning_rate: float = 1e-5,
    ) -> dict:
        """Fine-tune detector on custom dataset."""
        
        # Freeze backbone, train detection head
        for param in self.model.backbone.parameters():
            param.requires_grad = False
        
        # Training loop
        optimizer = torch.optim.AdamW(
            self.model.parameters(),
            lr=learning_rate,
        )
        
        for epoch in range(epochs):
            for batch in dataset.train_loader:
                outputs = self.model(**batch)
                loss = outputs.loss
                loss.backward()
                optimizer.step()
                optimizer.zero_grad()
            
            # Validation
            metrics = self.evaluate(dataset.val_loader)
            logger.info(f"Epoch {epoch}: mAP={metrics['mAP']:.3f}")
        
        # Save model
        self.model.save_pretrained(output_dir)
        
        return metrics

# training/llm_finetune.py
from peft import LoraConfig, get_peft_model

class LLMFineTuner:
    def __init__(self, base_model: str):
        self.model = AutoModelForCausalLM.from_pretrained(base_model)
        
        # Configure LoRA
        lora_config = LoraConfig(
            r=16,
            lora_alpha=32,
            target_modules=["q_proj", "v_proj"],
            lora_dropout=0.05,
        )
        self.model = get_peft_model(self.model, lora_config)
    
    def train(self, dataset: RiskScoringDataset) -> dict:
        """Fine-tune LLM on user's risk scoring preferences."""
        # User feedback: "This event was scored 70 but should be 30"
        # Train model to predict user's preferred scores
        ...
```

### Known Entity Matching

```python
# backend/services/known_entities.py
class KnownEntityMatcher:
    def __init__(self, embedding_model: str = "facenet"):
        self.model = load_embedding_model(embedding_model)
        self.threshold = 0.6  # Cosine similarity threshold
    
    async def match(
        self,
        detection: Detection,
        entity_type: str,
    ) -> KnownEntity | None:
        """Check if detection matches a known entity."""
        
        # Extract embedding
        embedding = await self.extract_embedding(detection.crop_path)
        
        # Get all known entities of this type
        entities = await self.repo.get_by_type(entity_type)
        
        # Find best match
        best_match = None
        best_score = 0
        
        for entity in entities:
            for stored_embedding in entity.embeddings:
                score = cosine_similarity(embedding, stored_embedding)
                if score > best_score and score > self.threshold:
                    best_score = score
                    best_match = entity
        
        return best_match
```

### Phase 1: Feedback Collection

- [ ] Create feedback data models
- [ ] Implement feedback API
- [ ] Add feedback buttons to frontend
- [ ] Store feedback with detection metadata

### Phase 2: Known Entity System

- [ ] Create KnownEntity model
- [ ] Implement embedding extraction
- [ ] Build matching service
- [ ] UI for managing known entities

### Phase 3: Dataset Building

- [ ] Implement DatasetBuilder
- [ ] COCO format export for detector
- [ ] Preference dataset for LLM
- [ ] Data augmentation

### Phase 4: Training Pipeline

- [ ] Detector fine-tuning script
- [ ] LLM LoRA fine-tuning script
- [ ] Training job management
- [ ] Model versioning

### Phase 5: Model Deployment

- [ ] Hot-swap trained models
- [ ] A/B testing framework
- [ ] Rollback capability
- [ ] Performance monitoring

## Acceptance Criteria

- [ ] Users can mark detections correct/incorrect
- [ ] Users can define known persons/vehicles
- [ ] Known entities reduce risk score
- [ ] Training can be triggered from UI
- [ ] Fine-tuned models improve accuracy
- [ ] Models can be rolled back if degraded
